clash-lib/src/Clash/Core/Type.hs

```diff
@@ -473,9 +471,17 @@ funSubst tcm (Just s) = uncurry go
     -- matching against `Char -> Char` we'd find a duplicate `a -> Char`. We
     -- can't think of any (type-checking) cases where these mappings would map
     -- to different types, so this is OK for our purposes.
-    go (tyView -> FunTy a b) (tyView -> FunTy a' b') =
-      (++) <$> funSubst tcm (Just s) (a, a')
-           <*> funSubst tcm (Just s) (b, b')
+    go (AppTy a1 r1) (AppTy a2 r2) = do
+      s1 <- funSubst tcm (Just s) (a1,a2)
+      funSubst tcm (Just s1) (r1,r2)
```

Comment:
Very minor nitpick, and probably not worth fixing but: could you please format tuples with spaces? It helps readability for larger tuples.